GenericTensor/Functions/SquareMatrixFactory.cs

```diff
@@ -4,17 +4,19 @@
 
 namespace GenericTensor.Functions
 {
-    internal static class SquareMatrixFactory<T>
+    internal static class SquareMatrixFactory<T> where T : notnull
     {
-        [ThreadStatic] internal static List<GenTensor<T>> TensorTempFactorySquareMatrices;
+        // [0] is 2x2 matrix, [1] is 3x3 matrix, etc.
+        static readonly List<GenTensor<T>> tensorTempFactorySquareMatrices = new List<GenTensor<T>>();
 
         internal static GenTensor<T> GetMatrix(int diagLength)
         {
-            if (TensorTempFactorySquareMatrices is null)
-                TensorTempFactorySquareMatrices = new List<GenTensor<T>> {null};
-            for (int i = TensorTempFactorySquareMatrices.Count; i <= diagLength; i++)
-                TensorTempFactorySquareMatrices.Add(new GenTensor<T>(i, i));
-            return TensorTempFactorySquareMatrices[diagLength - 1];
+            if (diagLength <= tensorTempFactorySquareMatrices.Count + 2)
+                lock (tensorTempFactorySquareMatrices)
+                    if (diagLength <= tensorTempFactorySquareMatrices.Count + 2)
```

Comment:
Those two lines are just the same? The one before and the one after lock().

Reply:
https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_C#